Android Custom Build Template fail > file path too long

<!-- Please search existing issues for potential duplicates before filing yours:
https://github.com/godotengine/godot/issues?q=is%3Aissue
-->

**Godot version:**
3.2.4.beta


**OS/device including version:**
Android 10 / Samsung Note 9

**Issue description:**

* What went wrong:

When exporting using Android custom build, the file path gets longer each time, until it's too long.
Deleting the Android/build folder in the Godot project and reexporting works fine.

What is causing the path to repeat like this?

On a sidenote, I am making custom export templates anyway, can I move the plugin from project/android/plugins to godot/modules and rebuild export templates to skip this whole android custom build template process?

Execution failed for task ':mergeDebugAssets'.
> A failure occurred while executing com.android.build.gradle.internal.tasks.Workers$ActionFacade
   > java.nio.file.FileSystemException: /Users/kakapoopie/HEAVYPAINT_source/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/release/out/android/build/build/intermediates/merged_assets/release/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out: File name too long


**Steps to reproduce:**


**Minimal reproduction project:**
<!-- A small Godot project which reproduces the issue. Drag and drop a zip archive to upload it. -->

Have tried with todays commit 3032b3835, still getting the same error.  Happens after about one click deploying about 10 times, the file path gets longer each subsequent deploy until it's too much.

    Caused by: java.nio.file.FileSystemException: /Users/kakapoopie/HEAVYPAINT_source/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/Icons: File name too long
    	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:91)
    	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
    	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
    	at sun.nio.fs.UnixFileSystemProvider.createDirectory(UnixFileSystemProvider.java:384)
    	at java.nio.file.Files.createDirectory(Files.java:674)
    	at java.nio.file.Files.createAndCheckIsDirectory(Files.java:781)
    	at java.nio.file.Files.createDirectories(Files.java:767)
    	at com.android.ide.common.resources.MergedAssetWriter$AssetWorkAction.run(MergedAssetWriter.java:84)
    	... 31 more
    

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':mergeDebugAssets'.
> A failure occurred while executing com.android.build.gradle.internal.tasks.Workers$ActionFacade
   > java.nio.file.FileSystemException: /Users/kakapoopie/HEAVYPAINT_source/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/android/build/build/intermediates/merged_assets/debug/out/Icons: File name too long

I've got a slightly different error with 3.2.4 Beta, similar conditions (happens after one click deploy a few times on android)

> Task :compressDebugAssets FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compressDebugAssets'.
> Multiple task action failures occurred:
   > A failure occurred while executing com.android.build.gradle.internal.tasks.CompressAssetsWorkAction
      > Java heap space
   > A failure occurred while executing com.android.build.gradle.internal.tasks.CompressAssetsWorkAction
      > Java heap space

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 1m 9s
25 actionable tasks: 2 executed, 23 up-to-date

@m4gr3d 
> Task :preBuild UP-TO-DATE
> Task :preDebugBuild UP-TO-DATE
> Task :compileDebugAidl NO-SOURCE
> Task :compileDebugRenderscript NO-SOURCE
> Task :generateDebugBuildConfig UP-TO-DATE
> Task :javaPreCompileDebug UP-TO-DATE
> Task :checkDebugAarMetadata UP-TO-DATE
> Task :generateDebugResValues UP-TO-DATE
> Task :generateDebugResources UP-TO-DATE

> Task :mergeDebugResources FAILED
/Users/kakapoopie/HEAVYPAINT_source/android/build/res/mipmap-xxxhdpi-v4/icon_foreground.png.import: Resource and asset merger: The file name must end with .xml or .png
/Users/kakapoopie/HEAVYPAINT_source/android/build/res/drawable/splash.png.import: Resource and asset merger: The file name must end with .xml or .png
/Users/kakapoopie/HEAVYPAINT_source/android/build/res/mipmap-hdpi-v4/icon_foreground.png.import: Resource and asset merger: The file name must end with .xml or .png
/Users/kakapoopie/HEAVYPAINT_source/android/build/res/mipmap-xhdpi-v4/icon_foreground.png.import: Resource and asset merger: The file name must end with .xml or .png
/Users/kakapoopie/HEAVYPAINT_source/android/build/res/mipmap-mdpi-v4/icon_foreground.png.import: Resource and asset merger: The file name must end with .xml or .png
/Users/kakapoopie/HEAVYPAINT_source/android/build/res/mipmap/icon_foreground.png.import: Resource and asset merger: The file name must end with .xml or .png
/Users/kakapoopie/HEAVYPAINT_source/android/build/res/mipmap-xxhdpi-v4/icon_foreground.png.import: Resource and asset merger: The file name must end with .xml or .png

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':mergeDebugResources'.
> /Users/kakapoopie/HEAVYPAINT_source/android/build/res/mipmap-xxxhdpi-v4/icon_foreground.png.import: Error: The file name must end with .xml or .png
  /Users/kakapoopie/HEAVYPAINT_source/android/build/res/drawable/splash.png.import: Error: The file name must end with .xml or .png
  /Users/kakapoopie/HEAVYPAINT_source/android/build/res/mipmap-hdpi-v4/icon_foreground.png.import: Error: The file name must end with .xml or .png
  /Users/kakapoopie/HEAVYPAINT_source/android/build/res/mipmap-xhdpi-v4/icon_foreground.png.import: Error: The file name must end with .xml or .png
  /Users/kakapoopie/HEAVYPAINT_source/android/build/res/mipmap-mdpi-v4/icon_foreground.png.import: Error: The file name must end with .xml or .png
  /Users/kakapoopie/HEAVYPAINT_source/android/build/res/mipmap/icon_foreground.png.import: Error: The file name must end with .xml or .png
  /Users/kakapoopie/HEAVYPAINT_source/android/build/res/mipmap-xxhdpi-v4/icon_foreground.png.import: Error: The file name must end with .xml or .png

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 1s
5 actionable tasks: 1 executed, 4 up-to-date

--- a/editor/editor_export.cpp
+++ b/editor/editor_export.cpp
@@ -507,6 +507,11 @@ void EditorExportPlatform::_edit_files_with_filter(DirAccess *da, const Vector<S
 		if (dir.begins_with(".")) {
 			continue;
 		}
+
+		if (EditorFileSystem::_should_skip_directory(cur_dir + dir)) {
+			continue;
+		}
+
 		da->change_dir(dir);
 		_edit_files_with_filter(da, p_filters, r_list, exclude);
 		da->change_dir("..");

--- a/editor/editor_file_system.cpp
+++ b/editor/editor_file_system.cpp
@@ -669,10 +669,7 @@ void EditorFileSystem::_scan_new_dir(EditorFileSystemDirectory *p_dir, DirAccess
 				continue;
 			}
 
-			if (FileAccess::exists(cd.plus_file(f).plus_file("project.godot"))) { // skip if another project inside this
-				continue;
-			}
-			if (FileAccess::exists(cd.plus_file(f).plus_file(".gdignore"))) { // skip if another project inside this
+			if (_should_skip_directory(cd.plus_file(f))) {
 				continue;
 			}
 
@@ -874,10 +871,7 @@ void EditorFileSystem::_scan_fs_changes(EditorFileSystemDirectory *p_dir, const 
 
 				int idx = p_dir->find_dir_index(f);
 				if (idx == -1) {
-					if (FileAccess::exists(cd.plus_file(f).plus_file("project.godot"))) { // skip if another project inside this
-						continue;
-					}
-					if (FileAccess::exists(cd.plus_file(f).plus_file(".gdignore"))) { // skip if another project inside this
+					if (_should_skip_directory(cd.plus_file(f))) {
 						continue;
 					}
 
@@ -1977,6 +1971,20 @@ Error EditorFileSystem::_resource_import(const String &p_path) {
 	singleton->reimport_files(files);
 
 	return OK;
+}
+
+bool EditorFileSystem::_should_skip_directory(const String &p_path) {
+	if (FileAccess::exists(p_path.plus_file("project.godot"))) {
+		// skip if another project inside this
+		return true;
+	}
+
+	if (FileAccess::exists(p_path.plus_file(".gdignore"))) {
+		// skip if a `.gdignore` file is inside this
+		return true;
+	}
+
+	return false;
 }
 
 bool EditorFileSystem::is_group_file(const String &p_path) const {

--- a/editor/editor_file_system.h
+++ b/editor/editor_file_system.h
@@ -262,6 +262,8 @@ public:
 	bool is_group_file(const String &p_path) const;
 	void move_group_file(const String &p_path, const String &p_new_path);
 
+	static bool _should_skip_directory(const String &p_path);
+
 	EditorFileSystem();
 	~EditorFileSystem();
 };

--- a/platform/android/export/export.cpp
+++ b/platform/android/export/export.cpp
@@ -2414,6 +2414,9 @@ public:
 		print_verbose("- custom build enabled: " + bool_to_string(use_custom_build));
 		print_verbose("- apk expansion enabled: " + bool_to_string(apk_expansion));
 		print_verbose("- enabled abis: " + String(",").join(enabled_abis));
+		print_verbose("- export filter: " + itos(p_preset->get_export_filter()));
+		print_verbose("- include filter: " + p_preset->get_include_filter());
+		print_verbose("- exclude filter: " + p_preset->get_exclude_filter());
 
 		Ref<Image> splash_image;
 		Ref<Image> splash_bg_color_image;

